test/e2e/node_feature_discovery.go

```diff
@@ -419,6 +423,103 @@ var _ = SIGDescribe("Node Feature Discovery", func() {
+
+				// Create CRDs
+				By("Creating NodeFeatureRule CRD")
+				var err error
+				crd, err = testutils.CreateNodeFeatureRulesCRD(extClient)
+				Expect(err).NotTo(HaveOccurred())
+			})
+
+			AfterEach(func() {
+				err := extClient.ApiextensionsV1().CustomResourceDefinitions().Delete(context.TODO(), crd.Name, metav1.DeleteOptions{})
+				Expect(err).NotTo(HaveOccurred())
+			})
+
+			It("custom labels should create labels should be created", func() {
+				By("Creating nfd-worker daemonset")
+				workerArgs := []string{"-feature-sources=fake", "-label-sources=", "-sleep-interval=1s"}
+				workerDS := testutils.NFDWorkerDaemonSet(fmt.Sprintf("%s:%s", *dockerRepo, *dockerTag), workerArgs)
+				workerDS, err := f.ClientSet.AppsV1().DaemonSets(f.Namespace.Name).Create(context.TODO(), workerDS, metav1.CreateOptions{})
+				Expect(err).NotTo(HaveOccurred())
+
+				By("Waiting for daemonset pods to be ready")
+				Expect(testutils.WaitForPodsReady(f.ClientSet, f.Namespace.Name, workerDS.Spec.Template.Labels["name"], 5)).NotTo(HaveOccurred())
+
+				expected := map[string]string{
+					"feature.node.kubernetes.io/e2e-flag-test-1":      "true",
+					"feature.node.kubernetes.io/e2e-attribute-test-1": "true",
+					"feature.node.kubernetes.io/e2e-instance-test-1":  "true"}
+
+				By("Creating NodeFeatureRules #1")
+				Expect(testutils.CreateNodeFeatureRuleFromFile(nfdClient, "nodefeaturerule-1.yaml")).NotTo(HaveOccurred())
+
+				By("Verifying node labels from NodeFeatureRules #1")
+				Expect(waitForNfdNodeLabels(f.ClientSet, expected)).NotTo(HaveOccurred())
+
+				By("Creating NodeFeatureRules #2")
+				Expect(testutils.CreateNodeFeatureRuleFromFile(nfdClient, "nodefeaturerule-2.yaml")).NotTo(HaveOccurred())
+
+				// Add features from NodeFeatureRule #2
+				expected["feature.node.kubernetes.io/e2e-matchany-test-1"] = "true"
+				expected["feature.node.kubernetes.io/e2e-template-test-1-instance_1"] = "found"
+				expected["feature.node.kubernetes.io/e2e-template-test-1-instance_2"] = "found"
 
+				By("Verifying node labels from NodeFeatureRules #1 and #2")
+				Expect(waitForNfdNodeLabels(f.ClientSet, expected)).NotTo(HaveOccurred())
+			})
+		})
+	})
 })
+
+// waitForNfdNodeLabels waits for node to be labeled as expected.
+func waitForNfdNodeLabels(cli clientset.Interface, expected map[string]string) error {
+	poll := func() error {
+		nodeList, err := cli.CoreV1().Nodes().List(context.TODO(), metav1.ListOptions{})
+		if err != nil {
+			return err
+		}
+		for _, node := range nodeList.Items {
+			labels := nfdLabels(node.Labels)
+			if !cmp.Equal(expected, labels) {
+				return fmt.Errorf("node %q labels do not match expected, diff (expected vs. received): %s", node.Name, cmp.Diff(expected, labels))
+			}
+		}
+		return nil
+	}
+
+	// Simple and stupid re-try loop
+	var err error
+	for retry := 0; retry < 3; retry++ {
+		if err = poll(); err == nil {
+			return nil
+		}
+		time.Sleep(2 * time.Second)
```

Comment:
If we are using default `sleepInterval` of 60 seconds, isn't the sleep interval of 2 seconds too short for the retry?

Reply:
The e2e tests (and this polling) is in no sync with the worker pods. There can basically be any random timeshift (deployment delays, image pulling etc) so we have have no idea when the labeling really happens. It might be just 1 second after we just polled and we don't want to wait for e.g. 60 seconds. I just picked some small number because we basically just want to see the desired state as fast as possible and continue the test sequence. Makes sense?

Reply:
Thanks for the explanation. This makes sense indeed.